### ISSUE TITLE
Using Class in Ingress Causes Error

### ISSUE THREAD

**Issue by treyhendon** [strimzi/strimzi-kafka-operator#4679]
**Describe the bug**
Trying to use nginx ingress and receiving error from the operator.

**To Reproduce**
1. Create a listener of type `ingress`
2. Add `configuration.class` value of `nginx`
3. Apply
4. See error in operator

```
2021-03-30 21:27:49 WARN AbstractOperator:508 - Reconciliation #32(watch) Kafka(strimzi/company): Failed to reconcile
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://10.0.0.1/apis/networking.k8s.io/v1/namespaces/strimzi/ingresses. Message: Ingress.extensions "company-kafka-external-bootstrap" is invalid: annotations.kubernetes.io/ingress.class: Invalid value: "nginx": can not be set when the class field is also set. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=annotations.kubernetes.io/ingress.class, message=Invalid value: "nginx": can not be set when the class field is also set, reason=FieldValueInvalid, additionalProperties={})], group=extensions, kind=Ingress, name=company-kafka-external-bootstrap, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Ingress.extensions "company-kafka-external-bootstrap" is invalid: annotations.kubernetes.io/ingress.class: Invalid value: "nginx": can not be set when the class field is also set, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:570) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:509) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:474) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:435) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:250) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:871) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:366) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:85) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.internalCreate(AbstractResourceOperator.java:213) ~[io.strimzi.operator-common-0.22.1.jar:0.22.1]
at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.lambda$reconcile$0(AbstractResourceOperator.java:104) ~[io.strimzi.operator-common-0.22.1.jar:0.22.1]
at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313) ~[io.vertx.vertx-core-3.9.1.jar:3.9.1]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty.netty-common-4.1.60.Final.jar:4.1.60.Final]
at java.lang.Thread.run(Thread.java:834) [?:?]
```

**Expected behavior**
The ingress should be created with the proper kube annotation for `ingressClassName`.

**Environment (please complete the following information):**
 - Strimzi version: 0.22.1
 - Installation method: Helm chart from https://strimzi.io/charts/
 - Kubernetes cluster: Kubernetes 1.19
 - Infrastructure: Azure AKS

**YAML files and logs**

```
      - name: external
        port: 9096
        type: ingress
        tls: true
        configuration:
          class: nginx
          brokerCertChainAndKey:
            secretName: kubeingress-tls
            certificate: tls.crt
            key: tls.key
          bootstrap:
            host: kafka-bootstrap.test.company.com
          brokers:
            - broker: 0
              host: kafka-0.test.company.com
            - broker: 1
              host: kafka-1.test.company.com
            - broker: 2
              host: kafka-2.test.company.com
        authentication:
          type: scram-sha-512
```

**Additional context**
I'd really love to use Istio (as standard Kube Ingress Class) rather than nginx, but appears to be very specifically coded for nginx.

**Comment by scholzj** [comment]
Interesting ... my cluster (Kube 1.20.4, pure Kubernetes) does not seem to do this. It creates the Ingress resource with both the old annotation `kubernetes.io/ingress.class: nginx` as well as with `.spec.ingressClassName` and nothing has problem with it. I wonder if this is something related to Kube 1.19 or if something else is throwing the error. Can you check the Kubernetes logs who is throwing this message? Or whether maybe some admission controller installed by some application (such as Istio or your Nginx installation) can be causing this? The [docs](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) do not seem to indicate that they would be exclusive. It would be good to figure this out before removing the annotation since that might break it for other users.

----

> I'd really love to use Istio (as standard Kube Ingress Class) rather than nginx, but appears to be very specifically coded for nginx.

Well, we use and test with Nginx because that is the _default_ Kubernetes Ingress implementation (or used to be ... it seems the competition is big these days). There is not much specially coded for Nginx. AFAIK the only special thing are the `nginx.ingress.kubernetes.io/ssl-passthrough` and `nginx.ingress.kubernetes.io/backend-protocol` annotations. But you can easily set your own annotations if Istio needs that. I do not think there is much else what would be Nginx specific.

Unfortunately, every Ingress controller is a bit different ... if you want to contribute support and tests for the Istio Ingress, you are more than welcome (for a start explaining what does Istio Ingress not like on the current Ingress resources created by Strimzi would be a good start). AFAIK, on our Slack channel there were some users using Istio Ingress with Strimzi in the past, so you might also check with them.

Long term, I hope that the new Gateway API which will replace Ingress API will be more standardised and make it easier to use different Ingress controllers. From the latest docs, it is unfortunately not completely clear what parts of that does Istio support today ().

**Comment by treyhendon** [comment]
I totally understand about Nginx and the history there, no worries at all!

Regarding the error I posted above, it does appear to be Kube or AKS version related.  I've just updated to AKS 1.20.2 and now my ingress has the annotation and the `.spec.ingressClassName` set on it.

Additionally, the Ingress objects now have a status with a posted IP address!  I think we're good to go.

It appears the takeaway is that Kube 1.20+ is needed for Strimzi 0.22 (or whichever version began the dual notations, maybe 0.20?)  Thanks so much for pointing me in the direction of checking the versions.

**Comment by scholzj** [comment]
The support for the new Ingress API is from 0.22. It is good to know that it works in 1.20 and doesn't work in 1.19. Maybe I can workaround it to fix this for 1.19. I will need to have a look. Thanks for providing the info.

**Comment by treyhendon** [comment]
Update - I was able to work-around this by using the annotation feature for the listener.

```
      - name: external
        port: 9096
        type: ingress
        tls: true
        configuration:
          bootstrap:
            host: kafka-bootstrap.{{ .Values.strimzi.tld }}
            annotations:
              kubernetes.io/ingress.class: istio
          brokers:
            - broker: 0
              host: kafka-0.test.company.com
              annotations:
                kubernetes.io/ingress.class: istio
            - broker: 1
              host: kafka-1.test.company.com
              annotations:
                kubernetes.io/ingress.class: istio
            - broker: 2
              host: kafka-2.test.company.com
              annotations:
                kubernetes.io/ingress.class: istio
        authentication:
          type: scram-sha-512
```

Additional note, I do not actually have a Kube IngressClass object created in the cluster.

**Comment by scholzj** [comment]
> Well, maybe I thought it was resolved too quickly... I just deleted the cluster and was rebuilding to test the process. I'm now getting the initial error about not using the annotation while providing the class...
> 
> ```
> 2021-03-31 17:16:52 WARN AbstractOperator:508 - Reconciliation #233(timer) Kafka(strimzi/company): Failed to reconcile
> io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://10.0.0.1/apis/networking.k8s.io/v1/namespaces/strimzi/ingresses. Message: Ingress.extensions "company-kafka-external-bootstrap" is invalid: annotations.kubernetes.io/ingress.class: Invalid value: "istio": can not be set when the class field is also set. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=annotations.kubernetes.io/ingress.class, message=Invalid value: "istio": can not be set when the class field is also set, reason=FieldValueInvalid, additionalProperties={})], group=extensions, kind=Ingress, name=company-kafka-external-bootstrap, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Ingress.extensions "company-kafka-external-bootstrap" is invalid: annotations.kubernetes.io/ingress.class: Invalid value: "istio": can not be set when the class field is also set, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
> at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:570) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:509) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:474) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:435) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:250) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:871) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:366) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:85) ~[io.fabric8.kubernetes-client-5.0.2.jar:?]
> at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.internalCreate(AbstractResourceOperator.java:213) ~[io.strimzi.operator-common-0.22.1.jar:0.22.1]
> at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.lambda$reconcile$0(AbstractResourceOperator.java:104) ~[io.strimzi.operator-common-0.22.1.jar:0.22.1]
> at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313) ~[io.vertx.vertx-core-3.9.1.jar:3.9.1]
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
> at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty.netty-common-4.1.60.Final.jar:4.1.60.Final]
> at java.lang.Thread.run(Thread.java:834) [?:?]
> ```

Wait, so ... do you know what configuration did you had when you got this error? Did you had the `class` field specified and set to `istio` at that point?

> Update - I was able to work-around this by using the annotation feature for the listener.
> 
> ```
>       - name: external
>         port: 9096
>         type: ingress
>         tls: true
>         configuration:
>           bootstrap:
>             host: kafka-bootstrap.{{ .Values.strimzi.tld }}
>             annotations:
>               kubernetes.io/ingress.class: istio
>           brokers:
>             - broker: 0
>               host: kafka-0.test.company.com
>               annotations:
>                 kubernetes.io/ingress.class: istio
>             - broker: 1
>               host: kafka-1.test.company.com
>               annotations:
>                 kubernetes.io/ingress.class: istio
>             - broker: 2
>               host: kafka-2.test.company.com
>               annotations:
>                 kubernetes.io/ingress.class: istio
>         authentication:
>           type: scram-sha-512
> ```
> 
> Additional note, I do not actually have a Kube IngressClass object created in the cluster.

This is a bit easier to understand why does it work like a workaround in your case. In this case, the `ingressClassName` will not be set. So only the annotation will be used. However, I still don't understand why did this worked fine for me yesterday (In my case the IngressClass also didn't exist).

**Comment by treyhendon** [comment]
Yeah, it's very bizarre.  I had an `IngressClass` named `istio` in Kube before running in the Kafka CRD with the `configuration.class: istio`.  I wonder if the time-frame where it worked was because my cluster was possibly still updating some nodes from 1.19 to 1.20?  No matter what I try today, setting class in the Strimzi Kafka CRD fails.

To test further this morning, I dropped back and started over.  Kube wise I have Istio and Nginx controllers installed at this point, and only one IngressClass defined as nginx.  I removed the Kafka config and ran it in with no class and no annotations in the Listeners.  The Ingress objects came up, but never were assigned IP addresses.  After 5 min of waiting, I removed the Kafka config and ensured everything was deleted.  I then added the annotation for `kubernetes.io/ingress.class: nginx` to the listeners and issued the config.  Now I have valid Ingress objects (with IP address from nginx).

After spending the rest of my day not being able to connect, I found a [question answered a couple of weeks back](https://github.com/strimzi/strimzi-kafka-operator/issues/4545) which brought great clarity.  There are 2 Nginx Ingress Controllers.  I have been using the wrong one.

I now have my cluster up and working with external ingress via Kube's Nginx Ingress Controller.

When the dust settles, my team does plan to return to the idea of Istio Ingress and what it will take to properly configure it to allow TLS Passthrough.  Unfortunately, that might be while.

Bottom line today - something is still weird with the `configuration.class` value in the Listener.  I would propose, maybe just have that set the `spec.ingressClassName` since there is already a great way to add any annotation to the Ingress object in each one's definition.

**Comment by scholzj** [comment]
Sorry, I was busy with something else.

It looks like I managed to reproduce it finally in my environment ... it seems that:
* When you modify an existing Kubernetes resource, Kubernetes does not care about having both the `kubernetes.io/ingress.class` annotation as well as `spec.ingressClassName`. 
* When you try to create a new Ingress with both specified, it complains.

So when you have the Kafka cluster deployed first without having the `class: abc` in the Kafka CR and then add it later, Kube is happy and doesn't care about it. But when you set the `class` in the Kafka CR right from the beginning, it will reject it.

I'm not completely happy about it ... but I guess removing the annotation and having people set it manually if needed is the only real option to go forward. I will open a PR.

**Comment by treyhendon** [comment]
No worries.  Your findings correlate with what I was seeing.  I could not create the resource with both, but if I created it with the annotation, I could manually edit the Ingress and add the `spec.ingressClassName`.

**Comment by scholzj** [comment]
I opened #4715 for this.
